community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java

```diff
@@ -19,61 +19,502 @@
+        {
+            @Override
+            protected StoreSingleLabelCursor create()
+            {
+                return new StoreSingleLabelCursor( this );
+            }
+        };
+        nodeRelationshipCursor = new InstanceCache<StoreNodeRelationshipCursor>()
+        {
+            @Override
+            protected StoreNodeRelationshipCursor create()
+            {
+                return new StoreNodeRelationshipCursor( new RelationshipRecord( -1 ),
+                        neoStore,
+                        new RelationshipGroupRecord( -1, -1 ), storeStatement, this );
+            }
+        };
+        singlePropertyCursor = new InstanceCache<StoreSinglePropertyCursor>()
+        {
+            @Override
+            protected StoreSinglePropertyCursor create()
+            {
+                return new StoreSinglePropertyCursor( neoStore.getPropertyStore(), this );
+            }
+        };
+        allPropertyCursor = new InstanceCache<StorePropertyCursor>()
+        {
+            @Override
+            protected StorePropertyCursor create()
+            {
+                return new StorePropertyCursor( neoStore.getPropertyStore(), this );
+            }
+        };
+
+    }
+
+    @Override
+    public NodeItem get()
+    {
+        return this;
     }
 
     @Override
-    public long getId()
+    public long id()
     {
         return nodeRecord.getId();
     }
 
     @Override
-    public LabelCursor labels()
+    public Cursor<LabelItem> labels()
+    {
+        return labelCursor.get().init( parseLabelsField( nodeRecord ).get( neoStore.getNodeStore() ) );
+    }
+
+    @Override
+    public Cursor<LabelItem> label( int labelId )
+    {
+        return singleLabelCursor.get().init( parseLabelsField( nodeRecord ).get( neoStore.getNodeStore() ), labelId );
+    }
+
+    @Override
+    public Cursor<PropertyItem> properties()
+    {
+        return allPropertyCursor.get().init( nodeRecord.getNextProp() );
+    }
+
+    @Override
+    public Cursor<PropertyItem> property( int propertyKeyId )
+    {
+        return singlePropertyCursor.get().init( nodeRecord.getNextProp(), propertyKeyId );
+    }
+
+    @Override
+    public Cursor<RelationshipItem> relationships( Direction direction )
+    {
+        return nodeRelationshipCursor.get().init( nodeRecord.isDense(), nodeRecord.getNextRel(), nodeRecord.getId(),
+                direction, null );
+    }
+
+    @Override
+    public Cursor<RelationshipItem> relationships( Direction direction, int... relTypes )
+    {
+        return nodeRelationshipCursor.get().init( nodeRecord.isDense(), nodeRecord.getNextRel(), nodeRecord.getId(),
+                direction, relTypes );
+    }
+
+    @Override
+    public Cursor<IntSupplier> relationshipTypes()
     {
-        return storeStatement.acquireLabelCursor( parseLabelsField( nodeRecord ).get( nodeStore ) );
+        if ( nodeRecord.isDense() )
+        {
+            return new Cursor<IntSupplier>()
```

Comment:
We go against the general pattern here and create a new Cursor for every call instead of using a single reusable instance, right? Is the thinking that this is a method that is called rarely and so doesn't warrant pre-instantiated and reusable cursors?